Improves v_total_vms

NickLaMuro opened the pull request:
Model code from https://github.com/ManageIQ/manageiq/pull/16438, updated to work in master (no difference in code, just different states of the code around it causing different diffs).  Performance info can be found in that PR.

Effectively this allows you to make a single query to services and get the info necessary to determine the `v_total_vms`.  So running:

```ruby
Service.select(Service.arel_table[Arel.star], :v_total_vms).first.v_total_vms
```

And have only 1 query executed, looking like the following:

```sql
SELECT 'services'.*, (
  SELECT COUNT("vms"."id")
  FROM "services" "v_total_vms_services"
  INNER JOIN "service_resources" ON "service_resources"."service_id" = "v_total_vms_services"."id"
         AND "service_resources"."resource_type" = 'VmOrTemplate'
  INNER JOIN "vms" ON "vms"."id" = "service_resources"."resource_id"
  WHERE (("v_total_vms_services"."id" = "services"."id" OR "v_total_vms_services"."ancestry" ILIKE CONCAT("services"."id", '/%'))
          OR "v_total_vms_services"."ancestry" = CAST("services"."id" AS VARCHAR))
) v_total_vms
FROM "services" ORDER BY "services"."id"
ASC LIMIT 1
```

In the previous implementation, it would have look more like this:

```
irb> Service.first.v_total_vms
Service Load (0.3ms) SELECT  "services".* FROM "services" ORDER BY "services"."id" ASC LIMIT $1  [["LIMIT", 1]]
Service Inst Including Associations (0.2ms - 1rows)
Service Load (0.6ms)  SELECT "services".* FROM "services" WHERE (("services"."ancestry" LIKE '20000000000001/%' ...
ServiceResource Inst Including Associations (0.0ms - 0rows)
VmOrTemplate Load (0.5ms)  SELECT "vms".* FROM "vms" WHERE "vms"."id" = XXXXXX
VmOrTemplate Inst Including Associations (0.0ms - 0rows)
ServiceResource Load (0.2ms)  SELECT "service_resources".* FROM "service_resources" WHERE...
ServiceResource Inst Including Associations (0.0ms - 0rows)
VmOrTemplate Load (0.5ms)  SELECT "vms".* FROM "vms" WHERE "vms"."id" = XXXXXX
VmOrTemplate Inst Including Associations (0.0ms - 0rows)
...
```

As noted in the PRs that implemented the base code for these features to work (#11502 and #13101), the nested SQL that get's generated because of these methods is a double edged sword.  When used with one record, the query is effectively harmless, but if used against the entirety of a large table, this could overload a DB as nested selects have to be evaluated against each record in the main query.  This is still quicker than the alternative, but too many of these types of virtual attributes in a select can be a problem.


Links
-----
* https://github.com/ManageIQ/manageiq/pull/16438
* Uses features from #11502 and #13101


Steps for Testing/QA
--------------------

Best to be tested in IRB with this change, as it isn't hooked up to the UI in most cases.

```ruby
irb> Service.select(Service.arel_table[Arel.star], :v_total_vms).map(&:v_total_vms)
```

With the above, you should only see one query (you might need to enable `toggle_console_sql_logging` if running in production mode).  Testing with a DB that has a large number of services would also be a good idea, as this has the potential to cause a lot of work on the DB, and ideally would be used in index based actions on the API.

NickLaMuro commented:
@miq-bot add-label performance

cc @abellotti

Fryguy commented on app/models/service.rb:
```diff
@@ -71,6 +70,9 @@ class Service < ApplicationRecord
   include_concern 'ResourceLinking'
 
+  virtual_total :v_total_vms, :vms,
+                :arel => aggregate_hardware_arel("v_total_vms", vms_tbl[:id].count, :skip_hardware => true)
```
I'm fine with this, but I'm curious why we would need an arel, when we already have virtual_total which is a specialized method to begin with.  Isn't it like specializing a specialized method?

NickLaMuro replied:
This would work if `:vms` was an actual relation on service, but it is not.  It is another `virtual_attribute`, which is an alias to another `virtual_attribute`.  forgetting some of the details, but there is also issues with the fact that this is a polymorphic relationship with VMs, and I don't think that `virtual_total` is built with that in mind.

NickLaMuro commented:
@Fryguy so, I think we have a problem...

Though there is some good news... I added some tests and the values are equivalent, so my code "works".

The problem is that the values returned in the tests probably don't make sense, and they probably have been broken for a bit.  This is not shown in the diff, but I am getting my test data from here:

https://github.com/NickLaMuro/manageiq/blob/fcb71ff/spec/models/service_spec.rb#L72-L91

But this is the important bits:

```ruby
@vm  = FactoryGirl.create(:vm_vmware)
@vm1 = FactoryGirl.create(:vm_vmware)
@vm2 = FactoryGirl.create(:vm_vmware)

@service    = FactoryGirl.create(:service)
@service_c1 = FactoryGirl.create(:service, :service => @service)
@service_c2 = FactoryGirl.create(:service, :service => @service_c1)
@service << @vm
@service_c1 << @vm1
@service_c2 << @vm1
@service_c2 << @vm2
```

Notice that there are only 3 VMs records being created, but there are 4 being returned.  This is because the counting is not accounting for uniq records, but I think it should be.

I don't know if the data sample is ever a real world scenario, but this would inflate the VM counts when this method is called (if it were to be used in reporting or something).  Not sure how to go about this.

Fryguy commented:
I don't think a VM can be in multiple services simultaneously in the real world.  @gmcculloug Can you comment here?

gmcculloug commented:
@Fryguy That's correct.  We enforce that here: [app/models/vm.rb#L36-L38](https://github.com/ManageIQ/manageiq/blob/dbaa1fc44c405d9b9c8ec8b0cc724b28da9d8fb9/app/models/vm.rb#L36-L38)

NickLaMuro commented:
@Fryguy @gmcculloug Thanks for the followup.  Seems like then this is something we can address at another time, since I do think regardless of model validation, we should try to make sure we are returning something that the user expects incase someone gets around that model validation or it gets removed at some point.

But as mentioned, doesn't have to happen now as this scenario isn't supposed to happen.

NickLaMuro commented:
@Fryguy just ran across this one that I still had open.  Is there any thing that is holding this one up?  I added some tests after a rebase, so I think I addressed your only concern you had after your initial review, and I think the concern that I brought up is a non-issue in the wild (just is weird in the tests otherwise).

NickLaMuro commented:
@Fryguy ping on above.

simaishi commented:
@NickLaMuro so... this goes with https://bugzilla.redhat.com/show_bug.cgi?id=1510469?  And since this PR is for master/Gaprindashvili and https://github.com/ManageIQ/manageiq/pull/16438 is for Euwe branch, how would Fine branch be fixed?

NickLaMuro commented:
@simaishi Sorry for the confusion, but let me explain:

tl; dr is you should just merge this, and only this, into 'G' for now and that should be ~~fine~~ okay (bad choice of words...).

So I originally put together https://github.com/ManageIQ/manageiq/pull/16438 together for a followup performance improvement to what Chris and Jason did in https://github.com/ManageIQ/manageiq/pull/16426, both of which were targeted fixes for EUWE and the BZ that you mentioned.  Mine was a improvement on what they were doing based on what I noticed separately and after it was mentioned (somewhere in the bowels of gitter) that https://github.com/ManageIQ/manageiq/pull/16426 was not enough of an improvement.  This was meant to be just a targeted fix for E only.

Per [a suggestion by Alberto](https://github.com/ManageIQ/manageiq/pull/16438#issuecomment-344323544) in that PR, we wanted to get the same performance on to master, which I agreed, and this PR was born.  I also suggested that it would probably make sense to split the effort into a part that fixes the models (which I did here), which could easily be applied all the way back to EUWE, and then a separate effort to get it into the API.  That second effort was more involved because of changes to the API controller code since EUWE and the API repo split.

* * *

All that said, I don't know where that ticket has landed at this point.  I was not the original owner of it, so I don't know if this even makes sense to backport to EUWE/FINE at this point.  I would ask @Fryguy to see if this effort is needed still on our older releases, other wise we can just get this into the newer releases and call it good.

I would need to also work up a PR to handled the controller code for master and G, and a separate PR for FINE and/or EUWE to allow the performance improvements to be leveraged in the API.  The API might be the same between E and F, so maybe we could backport from F to E.

simaishi commented:
Removing `gaprindashvili/yes` for now, as API change isn't ready.
